docs/proposals/001-in-place-upgrades.md

```diff
@@ -0,0 +1,257 @@
+<!--
+To start a new proposal, create a copy of this template on this directory and
+fill out the sections below.
+-->
+
+# Proposal information
+
+<!-- Index number -->
+- **Index**: 001
+
+<!-- Status -->
+- **Status**: **DRAFTING** <!-- **DRAFTING**/**ACCEPTED**/**REJECTED** -->
+
+<!-- Short description for the feature -->
+- **Name**: ClusterAPI In-Place Upgrades
+
+<!-- Owner name and github handle -->
+- **Owner**: Berkay Tekin Oz [@berkayoz](https://github.com/berkayoz) <!-- [@name](https://github.com/name) -->
+
+# Proposal Details
+
+## Summary
+<!--
+In a short paragraph, explain what the proposal is about and what problem
+it is attempting to solve.
+-->
+
+Canonical Kubernetes CAPI providers should reconcile workload clusters and perform in-place upgrades based on the metadata in the cluster manifest.
+
+This can be used in environments where rolling upgrades are not a viable option such as edge deployments and non-HA clusters.
+
+## Rationale
+<!--
+This section COULD be as short or as long as needed. In the appropriate amount
+of detail, you SHOULD explain how this proposal improves k8s-snap, what is the
+problem it is trying to solve and how this makes the user experience better.
+
+You can do this by describing user scenarios, and how this feature helps them.
+You can also provide examples of how this feature may be used.
+-->
+
+The current Cluster API implementation does not provide a way of updating machines in-place and instead follows a rolling upgrade strategy. 
+
+This means that a version upgrade would trigger a rolling upgrade, which is the process of creating new machines with desired configuration and removing older ones. This strategy is acceptable in most-cases for clusters that are provisioned on public or private clouds where having extra resources are not a concern.
+
+However this strategy is not viable for smaller bare-metal or edge deployments where resources are limited. This makes Cluster API not suitable out of the box for most of the use cases in industries like telco.
+
+We can enable the use of Cluster API in these use-cases by updating our providers to perform in-place upgrades.
+
+
+## User facing changes
+<!--
+This section MUST describe any user-facing changes that this feature brings, if
+any. If an API change is required, the affected endpoints MUST be mentioned. If
+the output of any k8s command changes, the difference MUST be mentioned, with a
+clear example of "before" and "after".
+-->
+
+Users will be able to perform in-place upgrades per machine basis by running:
+```sh
+kubectl annotate machine <machine-name> k8sd.io/in-place-upgrade-to={upgrade-option}
+```
+
+Users can also perform in-place upgrades on the entire cluster by running:
+```sh
+kubectl annotate cluster <cluster-name> k8sd.io/in-place-upgrade-to={upgrade-option}
```

Comment:
You want to upgrade the control-plane and the workers in separate steps.

Reply:
Good point, I'll update it with more details on this.

Reply:
@bschimke95 My approach for this would be to check which machines belong to the control plane, perform upgrades for those first then proceed with workers. We could wait for the upgrades on control plane machines to be completed before starting the worker node process. A possible `k8sd.io/in-place-upgrade-phase` with `control-plane` and `workers` annotation could control this. The cluster would get annotated with `k8sd.io/in-place-upgrade-phase` with `workers` once `control-plane` phase finishes.

Reply:
Sounds good. In your suggestion, would the user set `k8sd.io/in-place-upgrade-phase` or would this be set by the provider?

I read @ktsakalozos comment as "We need a way to let the user control this", right?

Reply:
This would be set by the provider and would be removed from the `Cluster` object once the upgrades are completed. 

For a more user controlled upgrade approach I feel it might be better/possible to go through `CK8sControlPlane` and `MachineDeployment` objects. It would be split more logically and users could annotate those in order as desired.

Reply:
Good point. How would this look like?
Cluster -> annotates CK8sControlPlane and MachineDeployment -> annotate respective machines? So annotating `Cluster` would basically be a convenience layer on the convenience layer.

Reply:
Correct, we would annotate child objects of the `Cluster`. This seems more balanced in terms of convenience and user control.

Reply:
I would move this to an "Out Of Scope" section, left as a future consideration for now. Just so that the focus can be on getting in-place upgrades to work properly, be well tested.

Then, we can return to build a nice orchestration around this.